cryptoki/src/context/mod.rs

```diff
@@ -44,12 +44,14 @@ pub(crate) struct Pkcs11Impl {
     // valid.
     _pkcs11_lib: cryptoki_sys::Pkcs11,
     pub(crate) function_list: cryptoki_sys::CK_FUNCTION_LIST,
+    pub(crate) function_list_30: Option<cryptoki_sys::CK_FUNCTION_LIST_3_0>,
```

Comment:
Would it be possible/better to have an enum here which can only be one of the two options and use `C_GetInfo` in the beginning of `new` to check which one the implementation is? That would be instead of trying to call `C_GetInterface`

Reply:
I think this wont work. The loaded shared object has only one symbol for PKCS#11 2.40 -- C_GetFunctionList. And the 3.0 has the two additional C_GetInterface and C_GetInterfaceList so we really can not call the C_GetInfo before getting the function list from either of the above functions. And if we call C_GetFunctionList unconditionally, the module will likely return a compatible function list with the version 2.40 in the C_GetInfo for compatibility (or at least some modules do that -- not sure though if it is the right thing to do).

Regarding the Enum, its probably a good suggestion. I initially did not want to break everything, but it might make the code cleaner if the macros and accessors will not get it more complicated. Let me try.

Reply:
Ah thank you for the explanation, I thought that `C_GetInfo` was also one of those accessible directly. Good to try with the enum but no need if it does not get simpler :)

Reply:
The enum might allow us to store the function pointer only once and just cast it to the correct type based on the interface version. But it might come later.